lightstep/sdk/metric/aggregator/histogram/histogram.go

```diff
@@ -0,0 +1,352 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package histogram // import "github.com/lightstep/otel-launcher-go/lightstep/sdk/metric/aggregator/histogram"
+
+import (
+	"fmt"
+	"sync"
+
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping"
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping/exponent"
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping/logarithm"
+
+	"github.com/lightstep/otel-launcher-go/lightstep/sdk/metric/aggregator"
+	"github.com/lightstep/otel-launcher-go/lightstep/sdk/metric/aggregator/aggregation"
+	"github.com/lightstep/otel-launcher-go/lightstep/sdk/metric/number"
+)
+
+var ErrNoSubtract = fmt.Errorf("histogram subtract not implemented")
+
+// Note: This code uses a Mutex to govern access to the exclusive
+// aggregator state.  This is in contrast to a lock-free approach
+// (as in the Go prometheus client) that was reverted here:
+// https://github.com/open-telemetry/opentelemetry-go/pull/669
+
+type (
+	// State observes counts observations in exponentially-spaced
+	// buckets.  It is configured with a maximum scale factor
+	// which determines resolution.  Scale is automatically
+	// adjusted to accommodate the range of input data.
+	State[N number.Any, Traits number.Traits[N]] struct {
+		lock    sync.Mutex
+		maxSize int32
+		data[N, Traits]
+	}
+
+	data[N number.Any, Traits number.Traits[N]] struct {
+		// sum is the sum of all Updates reflected in the
+		// aggregator.  It has the same type number as the
+		// corresponding sdkinstrument.Descriptor.
+		sum N
+		// count is incremented by 1 per Update.
+		count uint64
+		// zeroCount is incremented by 1 when the measured
+		// value is exactly 0.
+		zeroCount uint64
+		// positive holds the positive values
+		positive buckets
+		// positive holds the negative values in these buckets
+		// by their absolute value.
+		negative buckets
+		// mapping corresponds to the current scale, is shared
+		// by both positive and negative ranges.
+		mapping mapping.Mapping
+	}
+
+	Methods[N number.Any, Traits number.Traits[N], Storage State[N, Traits]] struct{}
```

Comment:
I wonder why do we need these `Methods` to be a separate structure that operates on `State` rather than just being methods of `State` in the first place?

This would make things simpler, e.g.:
```go
return newSyncView[
    N,
    histogram.State[N, Traits],
    histogram.Methods[N, Traits, histogram.State[N, Traits]], // this could be totally removed if methods were contained in the state.
](behavior)
```

Reply:
Good observation -- unfortunately it's difficult to make this idea work with Go generics without sacrificing something else. The problem is that the generic support can either know about a type or a pointer to a type, but it does not know there's a relationship between the pointer-to-type and the type itself. Methods{} is an empty struct that operates on pointers to the thing, whereas State{} is the thing that you can actually allocate inside another struct. The way the viewstate logic uses the aggregators (in both old SDK and this one) is that _two_ are allocated per synchronous accumulator. If you could find a way to work around the issue with generics, say with the use of another interface, you'd have more allocations when building a new allocator.

I take the appearance of empty-structs as the emergence of a new pattern for programming in Go with generics, basically, something new to get used to.

Reply:
Also: this needs to be documented. The same discussion was had in an OTel-Go meeting.